Source/Core/Jobs/CloseInactiveSessionsJob.cs

```diff
@@ -36,18 +39,24 @@ public class CloseInactiveSessionsJob : JobWithLockBase {
 
                 foreach (var sessionStart in results.Documents) {
                     var lastActivityUtc = sessionStart.Date.UtcDateTime.AddSeconds((double)sessionStart.Value.GetValueOrDefault());
-                    if (lastActivityUtc > inactivePeriod)
+                    var lastHeartbeatActivityUtc = await GetLastHeartbeatActivityUtcAsync(sessionStart).AnyContext();
+
+                    if (lastHeartbeatActivityUtc.HasValue && lastHeartbeatActivityUtc.Value > lastActivityUtc)
+                        sessionStart.UpdateSessionStart(lastHeartbeatActivityUtc.Value, isSessionEnd: lastHeartbeatActivityUtc.Value <= inactivePeriod);
+                    else if (lastActivityUtc <= inactivePeriod)
+                        sessionStart.UpdateSessionStart(lastActivityUtc, isSessionEnd: true);
+                    else
                         continue;
 
-                    sessionStart.UpdateSessionStart(lastActivityUtc, true);
                     sessionsToUpdate.Add(sessionStart);
-
                     Debug.Assert(sessionStart.Value != null && sessionStart.Value >= 0, "Session start value cannot be a negative number.");
                 }
 
-                if (sessionsToUpdate.Count > 0)
+                if (sessionsToUpdate.Count > 0) {
                     await _eventRepository.SaveAsync(sessionsToUpdate).AnyContext();
-                
+                    await _cacheClient.RemoveAllAsync(sessionsToUpdate.SelectMany(GetSessionHeartbeatCacheKeys)).AnyContext();
```

Comment:
This is the only real issue I see. The cache could be updated with a new timestamp and we are throwing it away. The window of time this happens is very very short..